Interaction between RSocket onClose and dispose

In 1.0.0-RC7 it seems like for `onClose()` to be subscribable, we have to do something with the "error" resulting from the `dispose()` that previously was not required. This is a _passing_ test written in `RSocketTest` that demonstrates what seems to be required:

```java
@Test
public void disposeAndOnClose() throws InterruptedException {
  CountDownLatch latch = new CountDownLatch(1);
  rule.crs.onClose()
          .map(v -> 1)
          .onErrorReturn(1) // I only care to know that it was closed, not why
          .subscribe(closed -> latch.countDown());
  rule.crs.dispose();

  assertThat(latch.await(1, TimeUnit.SECONDS)).isTrue();
}
```

Otherwise you get the exception:

```
reactor.core.Exceptions$ErrorCallbackNotImplemented: java.util.concurrent.CancellationException: Disposed
Caused by: java.util.concurrent.CancellationException: Disposed
	at io.rsocket.core.RSocketRequester.lambda$dispose$1(RSocketRequester.java:188)
	at io.rsocket.core.RSocketRequester.tryTerminate(RSocketRequester.java:690)
	at io.rsocket.core.RSocketRequester.dispose(RSocketRequester.java:188)
```

I did the weird `map(v -> 1).onErrorReturn(1)` thing because for the latch to count down we need a non-empty `Mono<Void>`, but paradoxically cannot create a `Mono<Void>` with null, the only valid value for `Void`. I'm probably just not seeing how to do this more idiomatically.

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
@@ -50,7 +50,6 @@ import io.rsocket.keepalive.KeepAliveHandler;
 import io.rsocket.keepalive.KeepAliveSupport;
 import io.rsocket.lease.RequesterLeaseHandler;
 import java.nio.channels.ClosedChannelException;
-import java.util.concurrent.CancellationException;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
 import java.util.function.Consumer;
@@ -64,6 +63,7 @@ import org.slf4j.LoggerFactory;
 import reactor.core.publisher.BaseSubscriber;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
 import reactor.core.publisher.Operators;
 import reactor.core.publisher.SignalType;
 import reactor.core.publisher.UnicastProcessor;
@@ -109,6 +109,7 @@ class RSocketRequester implements RSocket {
   private final RequesterLeaseHandler leaseHandler;
   private final ByteBufAllocator allocator;
   private final KeepAliveFramesAcceptor keepAliveFramesAcceptor;
+  private final MonoProcessor<Void> onClose;
   private final Scheduler serialScheduler;
 
   RSocketRequester(
@@ -129,6 +130,7 @@ class RSocketRequester implements RSocket {
     this.leaseHandler = leaseHandler;
     this.senders = new SynchronizedIntObjectHashMap<>();
     this.receivers = new SynchronizedIntObjectHashMap<>();
+    this.onClose = MonoProcessor.create();
     this.serialScheduler = serialScheduler;
 
     // DO NOT Change the order here. The Send processor must be subscribed to before receiving
@@ -182,17 +184,17 @@ class RSocketRequester implements RSocket {
 
   @Override
   public void dispose() {
-    tryTerminate(() -> new CancellationException("Disposed"));
+    tryShutdown();
   }
 
   @Override
   public boolean isDisposed() {
-    return connection.isDisposed();
+    return onClose.isDisposed();
   }
 
   @Override
   public Mono<Void> onClose() {
-    return connection.onClose();
+    return onClose;
   }
 
   private Mono<Void> handleFireAndForget(Payload payload) {
@@ -758,6 +760,11 @@ class RSocketRequester implements RSocket {
     senders.clear();
     receivers.clear();
     sendProcessor.dispose();
+    if (e == CLOSED_CHANNEL_EXCEPTION) {
+      onClose.onComplete();
+    } else {
+      onClose.onError(e);
+    }
   }
 
   private void handleSendProcessorError(Throwable t) {

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketRequester.java
@@ -64,7 +64,6 @@ import org.slf4j.LoggerFactory;
 import reactor.core.publisher.BaseSubscriber;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
-import reactor.core.publisher.MonoProcessor;
 import reactor.core.publisher.Operators;
 import reactor.core.publisher.SignalType;
 import reactor.core.publisher.UnicastProcessor;
@@ -110,7 +109,6 @@ class RSocketRequester implements RSocket {
   private final RequesterLeaseHandler leaseHandler;
   private final ByteBufAllocator allocator;
   private final KeepAliveFramesAcceptor keepAliveFramesAcceptor;
-  private final MonoProcessor<Void> onClose;
   private final Scheduler serialScheduler;
 
   RSocketRequester(
@@ -131,16 +129,12 @@ class RSocketRequester implements RSocket {
     this.leaseHandler = leaseHandler;
     this.senders = new SynchronizedIntObjectHashMap<>();
     this.receivers = new SynchronizedIntObjectHashMap<>();
-    this.onClose = MonoProcessor.create();
     this.serialScheduler = serialScheduler;
 
     // DO NOT Change the order here. The Send processor must be subscribed to before receiving
     this.sendProcessor = new UnboundedProcessor<>();
 
-    connection
-        .onClose()
-        .or(onClose)
-        .subscribe(null, this::tryTerminateOnConnectionError, this::tryTerminateOnConnectionClose);
+    connection.onClose().subscribe(null, this::tryTerminateOnConnectionError, this::tryShutdown);
     connection.send(sendProcessor).subscribe(null, this::handleSendProcessorError);
 
     connection.receive().subscribe(this::handleIncomingFrames, e -> {});
@@ -193,12 +187,12 @@ class RSocketRequester implements RSocket {
 
   @Override
   public boolean isDisposed() {
-    return onClose.isDisposed();
+    return connection.isDisposed();
   }
 
   @Override
   public Mono<Void> onClose() {
-    return onClose;
+    return connection.onClose();
   }
 
   private Mono<Void> handleFireAndForget(Payload payload) {
@@ -708,10 +702,6 @@ class RSocketRequester implements RSocket {
     tryTerminate(() -> e);
   }
 
-  private void tryTerminateOnConnectionClose() {
-    tryTerminate(() -> CLOSED_CHANNEL_EXCEPTION);
-  }
-
   private void tryTerminateOnZeroError(ByteBuf errorFrame) {
     tryTerminate(() -> Exceptions.from(0, errorFrame));
   }
@@ -721,6 +711,14 @@ class RSocketRequester implements RSocket {
       Throwable e = errorSupplier.get();
       if (TERMINATION_ERROR.compareAndSet(this, null, e)) {
         terminate(e);
+      }
+    }
+  }
+
+  private void tryShutdown() {
+    if (terminationError == null) {
+      if (TERMINATION_ERROR.compareAndSet(this, null, CLOSED_CHANNEL_EXCEPTION)) {
+        terminate(CLOSED_CHANNEL_EXCEPTION);
       }
     }
   }
@@ -760,7 +758,6 @@ class RSocketRequester implements RSocket {
     senders.clear();
     receivers.clear();
     sendProcessor.dispose();
-    onClose.onError(e);
   }
 
   private void handleSendProcessorError(Throwable t) {

--- a/rsocket-core/src/main/java/io/rsocket/core/RSocketResponder.java
+++ b/rsocket-core/src/main/java/io/rsocket/core/RSocketResponder.java
@@ -75,7 +75,6 @@ class RSocketResponder implements RSocket {
   private final PayloadDecoder payloadDecoder;
   private final ResponderLeaseHandler leaseHandler;
   private final Disposable leaseHandlerDisposable;
-  private final MonoProcessor<Void> onClose;
 
   private volatile Throwable terminationError;
   private static final AtomicReferenceFieldUpdater<RSocketResponder, Throwable> TERMINATION_ERROR =
@@ -110,7 +109,6 @@ class RSocketResponder implements RSocket {
     this.leaseHandler = leaseHandler;
     this.sendingSubscriptions = new SynchronizedIntObjectHashMap<>();
     this.channelProcessors = new SynchronizedIntObjectHashMap<>();
-    this.onClose = MonoProcessor.create();
 
     // DO NOT Change the order here. The Send processor must be subscribed to before receiving
     // connections
@@ -123,7 +121,6 @@ class RSocketResponder implements RSocket {
 
     this.connection
         .onClose()
-        .or(onClose)
         .subscribe(null, this::tryTerminateOnConnectionError, this::tryTerminateOnConnectionClose);
   }
 
@@ -256,12 +253,12 @@ class RSocketResponder implements RSocket {
 
   @Override
   public boolean isDisposed() {
-    return onClose.isDisposed();
+    return connection.isDisposed();
   }
 
   @Override
   public Mono<Void> onClose() {
-    return onClose;
+    return connection.onClose();
   }
 
   private void cleanup(Throwable e) {

--- a/rsocket-core/src/test/java/io/rsocket/core/RSocketTest.java
+++ b/rsocket-core/src/test/java/io/rsocket/core/RSocketTest.java
@@ -41,6 +41,8 @@ import org.junit.rules.ExternalResource;
 import org.junit.runner.Description;
 import org.junit.runners.model.Statement;
 import org.reactivestreams.Publisher;
+import reactor.core.Disposable;
+import reactor.core.Disposables;
 import reactor.core.publisher.DirectProcessor;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
@@ -50,6 +52,34 @@ import reactor.test.publisher.TestPublisher;
 public class RSocketTest {
 
   @Rule public final SocketRule rule = new SocketRule();
+
+  @Test
+  public void rsocketDisposalShouldEndupWithNoErrorsOnClose() {
+    RSocket requestHandlingRSocket =
+        new RSocket() {
+          final Disposable disposable = Disposables.single();
+
+          @Override
+          public void dispose() {
+            disposable.dispose();
+          }
+
+          @Override
+          public boolean isDisposed() {
+            return disposable.isDisposed();
+          }
+        };
+    rule.setRequestAcceptor(requestHandlingRSocket);
+    rule.crs
+        .onClose()
+        .as(StepVerifier::create)
+        .expectSubscription()
+        .then(rule.crs::dispose)
+        .expectComplete()
+        .verify(Duration.ofMillis(100));
+
+    Assertions.assertThat(requestHandlingRSocket.isDisposed()).isTrue();
+  }
 
   @Test(timeout = 2_000)
   public void testRequestReplyNoError() {
@@ -412,6 +442,9 @@ public class RSocketTest {
           new LocalDuplexConnection("server", allocator, clientProcessor, serverProcessor);
       LocalDuplexConnection clientConnection =
           new LocalDuplexConnection("client", allocator, serverProcessor, clientProcessor);
+
+      clientConnection.onClose().doFinally(__ -> serverConnection.dispose()).subscribe();
+      serverConnection.onClose().doFinally(__ -> clientConnection.dispose()).subscribe();
 
       requestAcceptor =
           null != requestAcceptor

